languages/python/oso/oso/enforcer.py

```diff
@@ -122,3 +122,62 @@ def authorized_actions(self, actor, resource, allow_wildcard=False) -> list:
             actions.add(action)
 
         return list(actions)
+
+    def authorize_field(self, actor, action, resource, field):
+        """Ensure that ``actor`` is allowed to perform ``action`` on a given
+        ``resource``'s ``field``.
+
+        If the action is permitted with an ``allow_field`` rule in the policy,
+        then this method returns ``None``. If the action is not permitted by the
+        policy, this method will raise a ``ForbiddenError`.
+
+        :param actor: The actor performing the request.
+        :param action: The action the actor is attempting to perform on the
+        field.
+        :param resource: The resource being accessed.
+        :param field: The name of the field being accessed.
+        """
+        if not self.policy.query_rule_once(
+            "allow_field", actor, action, resource, field
+        ):
+            raise self._get_error(False)
+
+    def authorized_fields(self, actor, action, resource, allow_wildcard=False) -> list:
```

Comment:
Would it make sense to combine this with an `authorize` check? (ditto for `authorize_field`).

I think the two would almost always be used together. First you check the action on the resource, then you check the fields.

Reply:
Hmm, I'm thinking about this. Not sure it'd be ergonomic. It makes some sense for `read`, but I feel like as a user I'd expect to explicitly perform both resource-level and field-level authorization on my own terms. I.e. they might happen in slightly different places, and the errors might be handled differently.

For example, I've been thinking about `authorize_field` method primarily being useful in a graphql setting, where you would have already resolved (and authorized) the parent resource.

And the `authorized_fields` would be useful for filtering the fields of an object, either coming in (POST parameters) or going out (JSON response). But again I'd expect the resource-level authorization to happen in a different place.

Also I think of field level authorization as `can this user perform this action on this field`, which doesn't _necessarily_ mean that the answer to `can this user perform this action on this resource` is true. In other words the "action space" of fields might be different than the "action space" of resources.

Does that make sense?

Reply:
> Hmm, I'm thinking about this. Not sure it'd be ergonomic. It makes some sense for `read`, but I feel like as a user I'd expect to explicitly perform both resource-level and field-level authorization on my own terms. I.e. they might happen in slightly different places, and the errors might be handled differently.
> 
> For example, I've been thinking about `authorize_field` method primarily being useful in a graphql setting, where you would have already resolved (and authorized) the parent resource.
> 
> And the `authorized_fields` would be useful for filtering the fields of an object, either coming in (POST parameters) or going out (JSON response). But again I'd expect the resource-level authorization to happen in a different place.

Yeah, I can definitely see the argument that it should be possible to split these two. 

> Also I think of field level authorization as `can this user perform this action on this field`, which doesn't _necessarily_ mean that the answer to `can this user perform this action on this resource` is true. In other words the "action space" of fields might be different than the "action space" of resources.

I don't agree with this. In my mind, having access to the resource should be a necessary prerequisite to having access to a field. You could suggest people do this in the policy:
```
allow_field(user, action, resource, field) if allow(user, action, resource) and ...
```

or through the enforcement API. Or by asking people to do the two authorize calls.

Depends on what we think the 80% case is... Should we optimise for the two being coupled? For example, an option would be making it `oso.authorize(user, "read", repo, fields=["name", "email"])` with fields as an optional param that does the two checks we spoke about. And `fields=None` is the default.

Pros/cons of that specific example: makes the API more complex, but it folds the best practice into one check.


Where I'm at is that we don't yet have enough knowledge on what the 80% case is here. So better to keep the two separate methods for now, and revisit later?

Reply:
Yeah good points -- I'd be curious to see more of how people end up using field-level enforcement. It's possible that `"read"` is in fact the 80% use-case and we should just optimize the methods around read-only access (making use of `enforcer.read_action`).

But for now, I think I share your hesitance to combine them into one method. How does this look otherwise?